x-pack/plugin/runtime-fields/qa/rest/src/yamlRestTest/java/org/elasticsearch/xpack/runtimefields/rest/CoreTestsWithRuntimeFieldsIT.java

```diff
@@ -0,0 +1,179 @@
+                        continue;
+                    }
+                    if ("false".equals(Objects.toString(propertyMap.get("index")))) {
+                        // If index is false we can't emulate with scripts
+                        continue;
+                    }
+                    if ("true".equals(Objects.toString(propertyMap.get("store")))) {
+                        // If store is true we can't emulate with scripts
+                        continue;
+                    }
+                    if (propertyMap.containsKey("ignore_above")) {
+                        // Scripts don't support ignore_above so we skip those fields
+                        continue;
+                    }
+                    if (propertyMap.containsKey("ignore_malformed")) {
+                        // Our source reading script doesn't emulate ignore_malformed
+                        continue;
+                    }
+                    String toLoad = painlessToLoadFromSource(name, type);
+                    if (toLoad == null) {
+                        continue;
+                    }
+                    propertyMap.put("type", "runtime_script");
+                    propertyMap.put("runtime_type", type);
+                    propertyMap.put("script", toLoad);
+                    propertyMap.remove("store");
+                    propertyMap.remove("index");
+                    propertyMap.remove("doc_values");
+                    include = true;
+                }
+            }
+        }
+        return include;
+    }
+
+    private static boolean containsUnsupportedSettings(Map<?, ?> settings) {
+        return
+        /*
+         * You can't sort on a runtime_keyword and it is hard to figure out
+         * if the sort was a runtime_keyword so lets just skip this test.
+         */
+        settings.containsKey("sort.field");
+    }
+
+    private static String painlessToLoadFromSource(String name, String type) {
+        String emit = PAINLESS_TO_EMIT.get(type);
+        if (emit == null) {
+            return null;
+        }
+        StringBuilder b = new StringBuilder();
+        b.append("def v = source['").append(name).append("'];\n");
+        b.append("if (v instanceof Iterable) {\n");
+        b.append("  for (def vv : ((Iterable) v)) {\n");
+        b.append("    if (vv != null) {\n");
+        b.append("      def value = vv;\n");
+        b.append("      ").append(emit).append("\n");
+        b.append("    }\n");
+        b.append("  }\n");
+        b.append("} else {\n");
+        b.append("  if (v != null) {\n");
+        b.append("    def value = v;\n");
+        b.append("    ").append(emit).append("\n");
+        b.append("  }\n");
+        b.append("}\n");
+        return b.toString();
+    }
+
+    private static final Map<String, String> PAINLESS_TO_EMIT = Map.ofEntries(
+        // TODO implement dates against the parser
+        Map.entry(
```

Comment:
This seems like the best approach for now but feels a bit fragile -- we're essentially duplicating the source parsing logic, which can be pretty nuanced. For example, for `boolean` fields, the value `""` is interpreted as `false`. We also don't cover cases where the value is malformed or missing from _source.

To make this more robust, I guess we could switch to 'source-only fields' when those are available?  That way we could reuse the logic from `fields` retrieval for parsing values from source. Or we could consider adding a way to access nicely-formatted source fields from painless?

Reply:
It really is pretty fragile. I was thinking of something similar to what you said - steal the source parsing and expose it to the script. That wouldn't be a "painless feature" so much as a feature of the script context of the runtime field. The advantage of this is that I know exactly which runtime field I'm running against so, for example, dates could use the the field's date parsing logic.